### PR TITLE
PR #2: Replacing the CAL/kg row of the (E)at menu with a more meaningful Satiety row.

### LINKED ISSUE #44365: [CR] Adjusting the new CAL/kg row on the (E)at menu to be CAL/L instead.
After I implemented the new CAL/kg row on the (E)at menu, worthwhile concerns were brought up that this does not effectively describe how filling a certain food is. It does, however, describe how energy dense a food is, which helps when hauling large amounts of food. This might be misleading, especially for new players, and is arguably less useful than a CAL/L row. 

Do you think the CAL/kg row should be changed to CAL/L?

### PR DESCRIPTION
**Pull request by gkarfakis19**
#### Summary

SUMMARY: Interface "Replaces the CAL/kg row of the (E)at menu with a more meaningful Satiety row."
#### Purpose of change

Continued from #44374, fixes #44365.

Uses actual consumption.cpp code to calculate the *exact* stomach fillingness value of each food, and compares this value with it's caloric content to accurately determine food quality. It then scales these results, color codes them, and displays them using a bar system on the (E)at menu, under "Satiety". This is a *far* less misleading result than the previous CAL/kg row I had added, and is a big QoL change aimed at new players and veterans who want to know exactly how valuable their food is.

#### Describe the solution

Edits game_inventory.cpp, and refactors consumption.cpp to use a new function in order to avoid repeating the same calculations twice.

#### Describe alternatives you've considered

Not changing anything.

#### Testing

Spawned in game. Values match up with mathematical expectations.

#### Additional context
"Vague" Values given.
image

As you can see, popcorn is now nearly worthless in terms of energy density, while glazed tenderloins are top notch food.

Originally, I had intended to show the exact values given certain character prerequisites, such as bionics, traits or skills. However,
there should be a broader conversation on whether these values should be shown at _all_. I will make a new PR after this one with the intended changes, so that this proposed feature can be discussed there.

### Comments Requested/ Additions that can be made:

Nourishment is aligned Left-to-Right. This can be changed if you think all values should be right to left aligned, but I think it looks better this way.
Create unit tests.

### CODE REVIEW

**Comment by gkarfakis19** [comment]
Ignore the commit-o-mania above, I forgot to squish the commits. 

Many people have argued *for* and *against* displaying the exact values calculated and not using a "vague " bar system. What I propose is that the values are vague by default, and only shown exactly should some prerequisites be met, such as bio_digestion, SELFAWARE, or high cooking skill. Additionally, even if those requirements are not fulfilled, it should give the exact numeric value if a screen reader is used. ( I think health also works that way?? Not sure. ). Nevertheless, changing the requirements is a trivial matter, as is disabling the whole vague bar thing altogether.

**Comment by wapcaplet** [comment]
@gkarfakis19 I took the liberty of writing up some basic tests of your `compute_effective_food_volume_ratio` function, which you can find in this commit: https://github.com/wapcaplet/Cataclysm-DDA/commit/ec27d286b869b941cdeb5626324c4e80b2d1a804 - Please feel free to cherry-pick that commit if you know how to do that, or I can open a pull request against your branch so you can simply merge it in. Let me know which you would prefer.

You can still squash the commits and do a "force push" to this new branch, if you like. The top answer on [this stackoverflow question](https://stackoverflow.com/questions/25356810/git-how-to-squash-all-commits-on-branch) shows how you can do it with the Git command-line. Afterward, a `git push --force-with-lease NewCalRowBranch` will replace all the "commit-o-mania" above. It's certainly not essential that you do this, but will make for a cleaner history.

Since your changes to `inventory_selector_preset` are kind of large, I suggest working toward refactoring some of that into another new helper function, perhaps one that takes the `item` and returns the colorized bars (and/or numbers). This would make that part easier to test as well, since we could add similar tests for how foods should be rendered, though I'm not saying we have to add tests for everything; those can always come in a future PR. But having a few tests often makes for quicker development, since it reduces the amount of manual in-game testing you have to do - especially if you end up using more complex conditions like the self-awareness trait, cooking skill, or bio-digestion, as you have noted in the comments.

I'm looking forward to seeing how this turns out!

**Comment by DoctorVanGogh** [comment]
>  I suggest working toward refactoring some of that into [...] helper function[s],

Yeah, I thought about doing that but wanted to wait until this gets merged, so this feature doesn't hang in limbo forever. Basically have something like `get_food_sat_bar(item)`, `compute_food_stats(player, item)` etc and (re)use those wherever appropriate.

**Comment by Qrox** [comment]
image

Why does crispy cranberry have different satiety in these two entries?

**Comment by DoctorVanGogh** [comment]
Looking at [old](https://user-images.githubusercontent.com/25616971/93723616-ab957e00-fba8-11ea-8e99-76cab14a9f04.png) [screenshots](https://user-images.githubusercontent.com/25616971/94073395-168cc200-fe00-11ea-9297-5c3aa98acc86.png) there may be a small bug in effective volume calculation in regard to stack size vs. consumed size.

**Comment by gkarfakis19** [comment]
I appreciate your feedback. I will incorporate all suggested changes into this PR (and force squish), but my internet access is quite flaky right now so it might take a day or two.

**Comment by Qrox** [comment]
> a discrepancy between found & self-squeezed cranberry juice

The calories column is displaying both items to have 105 calories though.

**Comment by gkarfakis19** [comment]
Oh boy, @wapcaplet was right in that following his instructions would squish my commits. However, it seems to have also pulled all PR's that I didn't have on my master, and it also requested a review automatically.

**Comment by gkarfakis19** [comment]
> 
> 
> image
> 
> Why does crispy cranberry have different satiety in these two entries?

I can reproduce this bug. Stacked cranberry juice has different satiety than non-stacked juice, but the amount of charges ( as long as it is above 1 ) doesn't seem to matter. Will investigate further.

File: src/game_inventory.cpp
Comment (wapcaplet) [inline src/game_inventory.cpp:492]: The `convert_volume` function is defined in `src/units_utility.h`, so you should be able to put this function in `src/consumption.cpp` as long as you have an `#include "units_utility.h"` at the top of that file with the other includes.

Keep in mind `convert_volume` returns a value based on the user's configured `VOLUME_UNITS` (in cups, liters, or quarts), which will affect the return value, in turn affecting how the bar-graph looks - not sure if that's what you intended. But now that you have factored it out into a separate function, this would be fairly easy to unit test.

File: src/output.cpp
Comment (wapcaplet) [inline src/output.cpp:1969]:
```suggestion
std::string satiety_bar( const int calpereffv )
```

The function does not actually display the bar, but only builds and returns the string, so it's probably better not to include the `display_` verb in the name.

File: src/output.h
Comment (wapcaplet) [inline src/output.h:331]:
```suggestion
std::string satiety_bar( int calpereffv );
```

I think this would be a better name (see related comment in the `.cpp` file).

**Review by wapcaplet — CHANGES_REQUESTED** [review]

**Comment by wapcaplet** [comment]
@gkarfakis19 Looking good so far! Having those calculations in smaller, self-contained functions will definitely make the code more readable and easier to change later. I will probably continue working on tests, now that those are modularized better.

To run the tests I provided, you can do something like `tests/cata_test effective_food_volume_ratio` (to match the name of the `TEST_CASE`). That works on Linux anyway; I don't know off the top of my head how to run them on other platforms, but someone on Discord can probably help if you aren't sure.

I don't know yet how you might solve the problem of numeric display for screen-readers; I am not even sure the existing health-bar system does that. Since the bars depend on the user's preferred volume units (cup, liter, or quart), the units may need to be worked into the table heading somehow, for the numeric version.

File: src/game_inventory.cpp
Comment (gkarfakis19) [reply src/game_inventory.cpp:492]: I originally did exactly as you said (actually VS did it for me, to be exact) and faced multiple build errors, so I gave up changing the include structure of the files. I will try again.

Comment (gkarfakis19) [reply src/game_inventory.cpp:492]: Actually, you are correct. I don't need to convert volumes, as the final result is so far removed from simple mathematical calculation it is essentially dimensionless anyway.

Comment (gkarfakis19) [reply src/game_inventory.cpp]: Issue is solved now, but I can once again confirm that #include in consumption.cpp gives 3 linking errors during build.

**Comment by DoctorVanGogh** [comment]
> I don't know yet how you might solve the problem of numeric display for screen-readers;

That's pretty simple to do (assuming you have some characters to make use of). You go down into the innermost `get_bar` method which does the string composition for `||||\` stuff and have it check for some (yet to be defined?) `screen_reader_substitute_numbers_for_bars` option.

If that is set, you don't return an aggregation of lines, but instead some textual value of `0-100%` (Hell, might still even be colorized). If you have less than 4 characters of width, it gets trickier. but you could still go with `1`, `.6`, ..., `0` in 2 characters or something like `+`, `9`, ..., `0` for 1 character.

**Comment by gkarfakis19** [comment]
I doubt that the travis CI fail has anything to do with my changes.

**Comment by wapcaplet** [comment]
> I doubt that the travis CI fail has anything to do with my changes.

Nope, looks unrelated.

As you probably saw, I submitted a PR to your branch with some more tests. Playtesting looks good too.

**Comment by gkarfakis19** [comment]
Travis CI fail has nothing to do with this PR.
Clang-tidy fail complains about using many dots instead of the unicode ellipsis, but using many dots is *required* in order to properly test the satiety bar, which displays with slashes and dots, *not* the ellipsis character.

**Comment by esotericist** [comment]
re: output from `get_bar()`:

> If you have less than 4 characters of width, it gets trickier.

fwiw, the vast majority of the cases where you would have less than 4 characters are circumstances where you're dealing with text which will naturally wrap, and thus it getting extended won't be terribly important.

i'm fairly certain that the cases where exact sizes are important are going to be situations where there's 5 characters, and if those don't cope with a shorter result that should be fixed at those callers (because they should be using strings with pad definitions anyway).

File: tests/comestible_test.cpp
Comment (esotericist) [inline tests/comestible_test.cpp:210]:
```suggestion
    // NOLINTNEXTLINE(cata-text-style): verbatim ellipses necessary for validation
    CHECK( satiety_bar( 396 ) == "<color_c_yellow>||</color>..." );
```

this should suppress the clang-tidy ellipses warning.

File: tests/comestible_test.cpp
Comment (esotericist) [inline tests/comestible_test.cpp:244]:
```suggestion
    // NOLINTNEXTLINE(cata-text-style): verbatim ellipses necessary for validation
    CHECK( satiety_bar( 0 ) == "<color_c_red></color>....." );
```

File: tests/comestible_test.cpp
Comment (esotericist) [inline tests/comestible_test.cpp:245]:
```suggestion
    // NOLINTNEXTLINE(cata-text-style): verbatim ellipses necessary for validation
    CHECK( satiety_bar( 100 ) == "<color_c_light_red>|</color>...." );
```

File: tests/comestible_test.cpp
Comment (esotericist) [inline tests/comestible_test.cpp:246]:
```suggestion
    // NOLINTNEXTLINE(cata-text-style): verbatim ellipses necessary for validation
    CHECK( satiety_bar( 200 ) == "<color_c_light_red>|\\</color>..." );
```

File: tests/comestible_test.cpp
Comment (esotericist) [inline tests/comestible_test.cpp:247]:
```suggestion
    // NOLINTNEXTLINE(cata-text-style): verbatim ellipses necessary for validation
    CHECK( satiety_bar( 300 ) == "<color_c_yellow>||</color>..." );
```

**Review by esotericist — COMMENTED** [review]
using the comment hints specified here should address those warnings.

**Pull request merged**
